clang/test/CIR/Transforms/simpl.cir

```diff
@@ -0,0 +1,41 @@
+// RUN: cir-opt %s -cir-simplify -o - | FileCheck %s
+
+!s32i = !cir.int<s, 32>
+!s64i = !cir.int<s, 64>
+module {
+cir.func @foo(%arg0: !cir.ptr<!s32i>) {
+    %0 = cir.alloca !cir.ptr<!s32i>, !cir.ptr<!cir.ptr<!s32i>>, ["ptr", init] {alignment = 8 : i64}
+    cir.store %arg0, %0 : !cir.ptr<!s32i>, !cir.ptr<!cir.ptr<!s32i>>
+    cir.scope {
+      %4 = cir.load %0 : !cir.ptr<!cir.ptr<!s32i>>, !cir.ptr<!s32i>
+      %5 = cir.const #cir.ptr<null> : !cir.ptr<!s32i>
+      %6 = cir.cmp(eq, %4, %5) : !cir.ptr<!s32i>, !s32i
+      %7 = cir.cast(int_to_bool, %6 : !s32i), !cir.bool
+      %8 = cir.unary(not, %7) : !cir.bool, !cir.bool
+      %9 = cir.cast(bool_to_int, %8 : !cir.bool), !s32i
+      %10 = cir.cast(int_to_bool, %9 : !s32i), !cir.bool
+      %11 = cir.unary(not, %10) : !cir.bool, !cir.bool
+      %12 = cir.cast(bool_to_int, %11 : !cir.bool), !s32i 
+      %13 = cir.cast(integral, %12 : !s32i), !s64i
```

Comment:
The opt logic seems not to take into account integer promotion happening with casts here, which could be unsafe. Depending on the pattern you are trying to get rid off, it might be better to make sure CIRGen gets it right out of the bat.